### ISSUE TITLE
Composition with a category-syntax-like template haskell preprocessor

### ISSUE THREAD

**Issue by dramforever** [mstksg/tensor-ops#3]
I think we could totally compose tensor operations by using a TH preprocessor like in [category-syntax](https://github.com/gelisam/category-syntax), and avoid the complexity manipulating stacks manually. The author of that library seems to be distracted to implementing categories the right way, but the whole idea is okay, AFAIK.

By using TH, we could transform this:

```
$(tensorOp [| do
  (x,y,z) <- getInput
  u <- f $ (x, z)
  v <- g $ (y, z)
  returnC <- (u, v)
|])
```

into something like

```
f >>> shuffling tensors >>> g >>> shuffling tensors
```

**Comment by mstksg** [comment]
That'd be great!  I'll look into TH processors like _category-syntax_ :)  I've also been looking into libraries like _syntactic_ too which can represent abstract typed functions and might have nice ways to build them.

I'm actually already running into some trouble with the stack-based system.  While it works pretty well for feed-forward networks, it's pretty cumbersome with recurrent neural networks that need to be unrolled to be "backpropagated through time".  Writing the stack manipulation for the unrolling is a bit annoying...fortunately, the tensor op only needs to be "unrolled" once, and the unrolled tensorop can be re-used to train multiple times.  But still, it'd be nice to get something that works more smoothly like in other libs!

**Comment by mstksg** [comment]
On the other hand, one of the reasons the stack-based composition was useful was that it made reverse-mode autodifferentiation practically trivial, and very performant.

If we switched to a more general abstract function representation, we might have to switch to using reflection-based methods and `unsafePerformIO`, like the _ad_ library does.

Oh, but I guess what you were proposing would also work if we kept the stack-based representation, but just offered a non-stack-based user API?  In that case, this point is moot.

**Comment by dramforever** [comment]
category-syntax is just a preprocessor, so it could easily integrate into the current interface. We still use the current `Shuffle`, `(>>>)`, etc., but we generate them instead of writing them manually.

**Comment by oliver-batchelor** [comment]
Hi guys,

From discussions I had with the author of category-syntax it seemed to me
like he was indicating it's quite a long way from finished. It also has the
downside that you can refer to the 'function' but not the value in the
network..

I'm not sure how tensor-ops handles parameters, maybe Justin @mstksg can
comment here - I like the idea that parameters are just a graph node, and
passed into the functions which is what I think tensor-ops does unless I'm
mistaken?

(unlike other deep learning frameworks where a function with parameters -
e.g. a linear layer 'contains' it's weights)

One important design factor is to separate the learning rules from the
gradient generation, in order to facilitate various learning algorithms,
e.g. SGD family, LBGFS etc.

The way this is handled in torch for example is that you can pull out all
the parameters into a single large vector (by making all the inner
parameter tensors 'views' into this large vector).

Other parameter related issues are implementing things like batch
normalization, which have parameters - but these parameters are updated as
a moving average, not via gradient descent.
https://arxiv.org/pdf/1502.03167.pdf

Maybe batch normalization (despite being used pretty much ubiquitously) is
just the flavor of the day, but I feel like there'll be other 'warts' which
will break the mold a little in future.

Other neural net implementations update these parameters as a side effect
of network evaluation, which is a dirty hack.

My idea here is to be able to tag parameter nodes, so they can be
extracted/shared as a group - e.g. so you can say, 'give me all the batch
normalization parameters'

I think a representation like syntactic would work - though I feel like
there are advantages to even more strongly typed representations e.g.
https://github.com/gelisam/strongly-typed-bound/blob/master/
StronglyTypedBound.hs

Where you can tell an expression has no free variables just from it's type
(i.e. the expression can be used as a NN)

I've also experimented with a graph representation (based on data-treify)
which seems reasonable as well, however I do have a distaste for the IO
based reification and feel like the lack of user control over sharing could
give unexpected results.. still, it's simple in comparison!

Just thought I'd put down my 2 cents here, in case anyone else has any
thoughts in the area!

Cheers,
Oliver

**Comment by mstksg** [comment]
@Saulzar

At this moment, the parameters are all indeed pulled into a single stack ("vector"), and running the network is considered an operation on this stack.  Composition of layers also composes the stack.

Referencing https://github.com/mstksg/tensor-ops/blob/dfb18a11aafccb4c5bb0054752e9ea2e3c365ac1/app/NeuralNets.hs#L61-L65 , a "neural network" operation contains:
1. `Prod t os`: A stack of parameters, where `os` contains the dimensions of every parameter
2. `TensorOp ('[i] ': os) '[ '[o] ]`: A tensor operation taking a single `t '[i]` (vector of `i` elements) and the parameters, and produces a `t '[o]` (simply vector of `o` elements)

To "run" the network, it's simply feeds the input vector consed with the stack of parameters https://github.com/mstksg/tensor-ops/blob/dfb18a11aafccb4c5bb0054752e9ea2e3c365ac1/app/NeuralNets.hs#L123-L128 : 

```
runTensorOp theTensorOp (inputVec :< stackOfParameters)
```

(`:<` is cons for `Prod`)

The code for "composing" networks is at https://github.com/mstksg/tensor-ops/blob/dfb18a11aafccb4c5bb0054752e9ea2e3c365ac1/app/NeuralNets.hs#L75-L83 -- it makes a newly new network by:
1. Concatenating the stacks of parameters, so the `Prod t os` and `Prod t ps` turns into `Prod t (os ++ ps)`
2. Composing the `TensorOp`s together, so the first one operates on the stack and leaves the output for the next one to take on.

So when the final network is run, the first "layer" does its stuff on the top of the stack and leaves a single output vector, which is exactly what the next "layer" is expecting.

In the end, the final network has a giant list of input parameters (`Prod t (h1 ++ h2 ++ h3 ++ ...)`), and the `TensorOp` in it chomps down each one on-by-one.

_Training_ the network is then simply asking for the gradient (using `gradTensorOp`) of an error function with respect to the inputs `Prod t ('[i] ': h1 ++ h2 ++ h3 ++ ...)`, and then adding this gradient to the internal list of parameters that the network contains.  (ignoring the head `'[i]` component of the gradient that we get as a side-effect)

Because stack access is O(1), there is no performance overhead for having large input stacks.  And, the stack-based model made the reverse-mode autodifferentation almost trivial to write/compose.

Currently not sure how the best way to "unroll" networks for backpropagation-through-time to work, though.  I know a couple of ways, but it's not clear what the "best" is.

I hope this clears up anything that wasn't clear about how parameters are treated and aggregated and trained for this system!  Let me know if you have any further questions.

**Comment by mstksg** [comment]
This all should be resolved now that the [backprop](http://hackage.haskell.org/package/backprop) library is released!  I will be rewriting this library to use that instead.
